src/core/smack.c

```diff
@@ -0,0 +1,953 @@
+    int num_trees = channels * rate_bytes;
+    hufftree8 *trees[4];
+    if (!read_audio_frame_trees(bs, trees, num_trees)) {
+        log_error("SMK: unable to read audio huffman trees", 0, 0);
+        return 0;
+    }
+
+    if (is_16bit) {
+        uint16_t *audio_data = (uint16_t *) s->frame_data.audio[track];
+        int index = 0;
+        for (int c = 0; c < channels; c++) {
+            // Base of channels is reversed
+            audio_data[channels - c - 1] = read_byte(bs) << 8 | read_byte(bs);
+        }
+        index = channels;
+
+        while (index < uncompressed_length / 2) {
+            for (int c = 0; c < channels; c++) {
+                uint16_t value = lookup_tree8(bs, trees[c * 2]) | lookup_tree8(bs, trees[c * 2 + 1]) << 8;
+                audio_data[index] = value + audio_data[index - channels];
+                index++;
+            }
+        }
+        s->frame_data.audio_len[track] = index * 2;
+    } else {
+        uint8_t *audio_data = s->frame_data.audio[track];
+        int index = 0;
+        for (int c = 0; c < channels; c++) {
+            // Base of channels is reversed
+            audio_data[channels - c - 1] = read_byte(bs);
+        }
+        index = channels;
+
+        while (index < uncompressed_length) {
+            for (int c = 0; c < channels; c++) {
+                audio_data[index] = lookup_tree8(bs, trees[c]) + audio_data[index - channels];
+                index++;
+            }
+        }
+        s->frame_data.audio_len[track] = index;
+    }
+    return 1;
+}
+
+static int decode_palette(smk s, uint8_t *data, int length)
+{
+    color_t new_palette[MAX_PALETTE];
+    int index = 0;
+    int color_index = 0;
+    while (index < length && color_index < MAX_PALETTE) {
+        if (data[index] & 0x80) {
+            // Copy from same position in previous palette
+            int num_entries = 1 + (data[index] & 0x7f);
+            if (num_entries + color_index > MAX_PALETTE) {
+                log_error("SMK: invalid palette data", 0, 0);
+                return 0;
+            }
+            memcpy(&new_palette[color_index], &s->frame_data.palette[color_index], sizeof(int32_t) * num_entries);
+            color_index += num_entries;
+            index++;
+        } else if (data[index] & 0x40) {
+            // Copy from 'offset' position in previous palette
+            int num_entries = 1 + (data[index] & 0x3f);
+            int offset = data[index + 1];
+            if (num_entries + color_index > MAX_PALETTE || num_entries + offset > MAX_PALETTE) {
+                log_error("SMK: invalid palette data", 0, 0);
+                return 0;
+            }
+            memcpy(&new_palette[color_index], &s->frame_data.palette[offset], sizeof(int32_t) * num_entries);
+            color_index += num_entries;
+            index += 2;
+        } else {
+            // Literal color
+            new_palette[color_index] =
+                    (PALETTE_MAP[data[index] & 0x3f] << 16) |
```

Comment:
Although more verbose, the `& 0x3F` is not strictly needed. If we ever get to this execution branch, the byte is already `00xxxxxx`.

Reply:
True, but you have to do it for the next two bytes anyway, so I'd rather keep the operation the same for all three bytes.

Reply:
Fair enough. Looks better anyway.